presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ProjectionPushThroughJoin.java

```diff
@@ -0,0 +1,182 @@
+        {
+            if (node.getSource() instanceof JoinNode) {
+                JoinNode joinNode = (JoinNode) node.getSource();
+                List<Symbol> leftOutputSymbols = joinNode.getLeft().getOutputSymbols();
+                List<Symbol> rightOutputSymbols = joinNode.getRight().getOutputSymbols();
+                Map<Symbol, Expression> leftAssignment = new HashMap<>();
+                Map<Symbol, Expression> rightAssignment = new HashMap<>();
+
+                // pushdown an expression through join
+                // if and only if the expression's dependency fields are all from one join child
+                // and do not exist in the other join child
+                for (Map.Entry<Symbol, Expression> entry : node.getAssignments().entrySet()) {
+                    Symbol key = entry.getKey();
+                    Expression expression = entry.getValue();
+                    if (expression instanceof SymbolReference) {
+                        if (((SymbolReference) expression).equals(key.toSymbolReference())) {
+                            continue;
+                        }
+                    }
+                    Set<Symbol> fields = DependencyExtractor.extractUnique(expression);
+                    if (fields.isEmpty()) {
+                        continue;
+                    }
+
+                    if (leftOutputSymbols.containsAll(fields) &&
+                        !rightOutputSymbols.stream().anyMatch(s -> fields.contains(s))) {
+                        leftAssignment.put(key, expression);
+                    }
+                    if (rightOutputSymbols.containsAll(fields) &&
+                        !leftOutputSymbols.stream().anyMatch(s -> fields.contains(s))) {
+                        rightAssignment.put(key, expression);
+                    }
+                }
+                if (!leftAssignment.isEmpty() || !rightAssignment.isEmpty()) {
+                    return context.rewrite(node.getSource(), Optional.of(new JoinPushdownAssignments(leftAssignment, rightAssignment)));
+                }
+                return node;
+            }
+            return node;
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Optional<JoinPushdownAssignments>> context)
+        {
+            Optional<JoinPushdownAssignments> assignment = context.get();
+            if (assignment == null || !assignment.isPresent()) {
+                return context.defaultRewrite(node, Optional.empty());
+            }
+            Map<Symbol, Expression> leftAssignment = assignment.get().getLeftAssignment();
+            Map<Symbol, Expression> rightAssignment = assignment.get().getRightAssignment();
+            List<Symbol> leftSymbols = node.getCriteria().stream().map(EquiJoinClause::getLeft).collect(toList());
+            List<Symbol> rightSymbols = node.getCriteria().stream().map(EquiJoinClause::getRight).collect(toList());
+            PlanNode left = node.getLeft();
```

Comment:
This should process node.getLeft/getRight recursively, otherwise, the transformation will stop at the first join it encounters.